### PR TITLE
Set precision=16 when use_amp is passed as True

### LINKED ISSUE #1143: `use_amp` is broken in 0.7.0
I see that `use_amp` is deprecated but since the Trainer still accepts it as an argument, I believe this is still a bug. 

The issue is that the [hook that deals with this](https://github.com/PyTorchLightning/pytorch-lightning/blob/3c2fd560aa4d31b4f48ee225b83361deec53d9c7/pytorch_lightning/core/hooks.py#L136-L143) specifically checks if `trainer.precision == 16`. However, the Trainer does not set `precision = 16` if `use_amp = True`.

Proposed solution:

if `use_amp` is passed in, set precision = 16 and raise a deprecation warning.

### PR DESCRIPTION
**Pull request by rmrao**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?   
- [x] Did you write any new necessary tests?  
- [x] If you made a notable change (that affects users), did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CHANGELOG.md)?

## What does this PR do?
Fixes #1143 

## PR review    
Anyone in the community is free to review the PR once the tests have passed.     
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -25,7 +25,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Fixed
 
--
+- Fixes `use_amp` issue ([#1145](https://github.com/PyTorchLightning/pytorch-lightning/pull/1145))
```
Comment (Borda) [inline CHANGELOG.md]:
```suggestion
- Fixes using deprecated `use_amp` attribute ([#1145](https://github.com/PyTorchLightning/pytorch-lightning/pull/1145))
```

File: pytorch_lightning/trainer/trainer.py [hunk 2]
```diff
@@ -414,6 +414,10 @@ def __init__(
                                        test_percent_check, overfit_pct)
 
         # 16 bit mixed precision training using apex
+        if use_amp:
+            warnings.warn("`use_amp` has been deprecated in favor of `precision` since v0.7.0"
+                          " and will be removed in v0.9.0", DeprecationWarning)
+            precision = 16
```
Comment (Borda) [inline pytorch_lightning/trainer/trainer.py]: rather add in Deprecated mixin setter like https://github.com/PyTorchLightning/pytorch-lightning/blob/3c2fd560aa4d31b4f48ee225b83361deec53d9c7/pytorch_lightning/trainer/deprecated_api.py#L75-L89

**Comment by rmrao** [comment]
Made the requested changes as far as I can tell. If I missed something, let me know!

**Comment by rmrao** [comment]
Hmm - it looks like trainer sets use_amp at some point and maybe this causes conflicts. I will debug and test locally.

**Comment by rmrao** [comment]
Ok, I have found and fixed the previous issue and run tests locally, so I believe things should work.

One concern is that it looks like the trainer uses `self.use_amp` in a number of places. The deprecation strategy will cause this to raise deprecation warnings though, even if the user passes in precision=16. So I see two options:

1. Change all checks from `self.use_amp` to `self.precision == 16` or maybe `self.precision < 32`.
2. Have the `use_amp` argument be deprecated, but still have the trainer set the `use_amp` attribute internally, without deprecated api (basically revert back to original solution)

Do you have a preference / other suggestions?

**Comment by Borda** [comment]
in such case, it would make sense to have `use_amp` as a standard property and make deprecated jut the setter to it... as we want to reduce this kind duplicated setting via `use_amp` and `precision`

**Comment by rmrao** [comment]
Right now, the trainer has this line

```python
if self.precision == 16 and self.num_tpu_cores is None:
    use_amp = True
self.init_amp(use_amp)
```
Then in `init_amp` we have

```python
def init_amp(self, use_amp):
     self.use_amp = use_amp and APEX_AVAILABLE
```

That means that the trainer, by default, accesses the setter for `use_amp`. The easiest way I see to solve this is to simply set `precision = 16` when `use_amp` is passed in as an argument, and avoid the setter/getter property. Alternatively, the property should read something like

```python
@property
def use_amp(self):
    return self.precision < 32 and self.num_tpu_cores is None and APEX_AVAILABLE

@use_amp.setter
def use_amp(self, use_amp):
    warnings.warn("`use_amp` has been deprecated in favor of `precision` since v0.7.0"
                             " and will be removed in v0.9.0", DeprecationWarning)
    self.precision = 16 if use_amp else 32
```

But I'm not sure what to do about `init_amp` in that case. Is this is a part of the API that can't / shouldn't be changed?

File: pytorch_lightning/trainer/deprecated_api.py [hunk 3]
```diff
@@ -87,3 +87,20 @@ def nb_sanity_val_steps(self, nb):
                       "`num_sanity_val_steps` since v0.5.0"
                       " and this method will be removed in v0.8.0", DeprecationWarning)
         self.num_sanity_val_steps = nb
+
+    @property
+    def use_amp(self):
+        """Back compatibility, will be removed in v0.9.0"""
+        warnings.warn("`use_amp` has been deprecated in favor of `precision` since v0.7.0"
+                      " and will be removed in v0.9.0", DeprecationWarning)
```
Comment (Borda) [inline pytorch_lightning/trainer/deprecated_api.py]:
```suggestion
```
as you mentioned this is used elsewhere...
also, move it to `Trainer`, not deprecated API

File: pytorch_lightning/trainer/deprecated_api.py [hunk 4]
```diff
@@ -87,3 +87,20 @@ def nb_sanity_val_steps(self, nb):
...
+
+    @use_amp.setter
+    def use_amp(self, use_amp):
+        """Back compatibility, will be removed in v0.9.0"""
+        warnings.warn("`use_amp` has been deprecated in favor of `precision` since v0.7.0"
+                      " and will be removed in v0.9.0", DeprecationWarning)
+        if use_amp:
+            self.precision = 16
+        else:
+            self.precision = 32
```
Comment (Borda) [inline pytorch_lightning/trainer/deprecated_api.py]:
```suggestion
        self.precision = 16 if use_amp else 32
```

File: pytorch_lightning/trainer/deprecated_api.py [hunk 5]
```diff
@@ -87,3 +87,20 @@ def nb_sanity_val_steps(self, nb):
...
+
+    @property
+    def use_amp(self):
+        """Back compatibility, will be removed in v0.9.0"""
+        warnings.warn("`use_amp` has been deprecated in favor of `precision` since v0.7.0"
+                      " and will be removed in v0.9.0", DeprecationWarning)
+        return self.precision == 16
+
+    @use_amp.setter
+    def use_amp(self, use_amp):
```
Comment (Borda) [inline pytorch_lightning/trainer/deprecated_api.py]: as this is deprecated until v0.9 and the class is v0.8 you need to create a new mixin

Comment (rmrao) [reply pytorch_lightning/trainer/deprecated_api.py]: Ah yes of course. Should this be in a new file, or in the same `deprecated_api.py` file?

**Comment by rmrao** [comment]
Can make those changes. What should I do about `init_amp`? Right now it looks like this:

```python
    def init_amp(self, use_amp):
        self.use_amp = use_amp and APEX_AVAILABLE
        if self.use_amp:
            log.info('Using 16bit precision.')


        if use_amp and not APEX_AVAILABLE:  # pragma: no cover
            msg = """
            You set `use_amp=True` but do not have apex installed.
            Install apex first using this guide and rerun with use_amp=True:
            https://github.com/NVIDIA/apex#linux

            this run will NOT use 16 bit precision
            """
            raise ModuleNotFoundError(msg)
```

That means that the trainer access the `self.use_amp` setter *regardless* of whether the user actually passes in the `use_amp` parameter. I can change the method signature, but this may cause problems for users if they have overridden this method. I am not sure if overriding this method is supported, but if it is, then I assume the signature needs to be kept the same.

One option is to make the following change:

```python
    def init_amp(self, precision):
        if isinstance(precision, bool):
              # raise deprecation warning
              self.use_amp = precision

        if self.use_amp:
            log.info('Using 16bit precision.')
```

File: pytorch_lightning/trainer/deprecated_api.py (hunk 5 above)
Comment (Borda) [reply pytorch_lightning/trainer/deprecated_api.py]: just a new class in the file is fine :]

**Comment by Borda** [comment]
this shall be fixed in #1163

**Comment by Borda** [comment]
@rmrao could you pls update from master?

**Comment by Borda** [comment]
@rmrao could you pls update from master? we have fixed the failure in #1163

**Comment by Borda** [comment]
@rmrao I have applied all previous suggestions, but we still need to as test for this one

**Comment by Borda** [comment]
@rmrao how is it going, could I help?

**Comment by williamFalcon** [comment]
@Borda can we merge this?

**Comment by Borda** [comment]
it seems that there is incompatibility with master... @rmrao ^^

**Comment by rmrao** [comment]
Hey - sorry last couple days have been super busy and I haven't had a chance to look at this closely. I'll try to get to it tomorrow.

**Comment by rmrao** [comment]
So I think some changes were made that I don't quite understand, and there are a few issues. As I see it there are two approaches:

1. Just set `precision` based on `use_amp` in `Trainer`'s `__init__`. Add a comment that this is deprecated. This has the advantage of very minimal code changes (in particular, no changes to `auto_mix_precision.py`). It has the disadvantage that it does not follow the pattern in `deprecated_api.py`

2. Add a `use_amp` property as a deprecated mixin. This has the advantage of following the pattern in `deprecated_api.py` but makes some complicated changes. First, python doesn't seem to support adding a getter in one class and a setter in another class, so we'd have to add both getter and setter to the deprecated API and add a note that the getter should be moved to `Trainer.py` in version 0.9.0. Second, `auto_mix_precision.py` actually sets the value of `self.use_amp` so it does access the setter. That means the setter can't actually be deprecated, or we have to find some other way of setting `use_amp`. Plausibly this should be changed so that `auto_mix_precision` sets the `precision` attribute, rather than the `use_amp` attribute.

Let me know which design spec you would like. 1. is simpler, but relies on a comment about deprecation rather than a mixin, 2. is more complex but does things as a mixin.

**Comment by williamFalcon** [comment]
@rmrao precision should only be set now through precision=xxx
use_amp is deprecated.

until deprepaction, we should just map use_amp to precision =16

**Comment by rmrao** [comment]
I understand - the question I had was how to do that. My initial commit did it by adding the following inside of Trainer's `__init__`:

```python
# TODO: use_amp is deprecated and will be removed in v0.9.0
if use_amp is not None:
    warnings.warn("The use_amp argument is deprecated and will be removed in v0.9.0", DeprecationWarning)
    self.precision = 16 if use_amp else 32
```

@Borda suggested doing it via a deprecated mixin, with a `use_amp` property w/ getter and setter. The problem with this approach is that the Trainer actually sets `self.use_amp = use_amp and APEX_AVAILABLE` inside of of the auto_mix_precision mixin. This would require changing the way that mixin works, if we want to turn `use_amp` into a property and make it altogether deprecated.

My question is which version is best. I am happy to revert to the original version for simplicity and speed. If you would prefer the deprecated mixin property, then I need to know how to change auto_mix_precision and every other point where Trainer accesses the `use_amp` property.

**Comment by Borda** [comment]
there was a point from @rmrao that we are still using the `use_amp` in the code still... :]

**Comment by Borda** [comment]
the setter was just because of the deprecation so whenever someone tries to change the AMP via this variable he gets the warning... the getter is there because `use_amp` was called elsewhere in code and for back-compatibility reasons

**Comment by williamFalcon** [comment]
@Borda @rmrao can we finish this PR? the 0.7.2 release is almost ready

**Comment by rmrao** [comment]
Ok I just went back to the easy solution. Tests passed locally, so should be good.

**Comment by Borda** [comment]
@rmrao it seems that I do have permission to push to your master to edit this PR, do you have enabled edit for maintained?

**Comment by rmrao** [comment]
I definitely have edit access for maintained enabled, just double checked.

**Comment by Borda** [comment]
just curious, there is also `use_amp` in the model... it is kind of duplication @jeremyjordan?

**Comment by Borda** [comment]
@rmrao did some more cleaning, could you please check and resolve conflicts...

**Comment by rmrao** [comment]
It looks good to me!

**Comment by williamFalcon** [comment]
@rmrao mind rebasing?

**Comment by Borda** [comment]
@williamFalcon rebased ^^

File: CHANGELOG.md [hunk 6]
```diff
@@ -81,6 +81,8 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Fixed checkpointing interval ([#1272](https://github.com/PyTorchLightning/pytorch-lightning/pull/1272)) 
 - Fixed validation and training loops run the partial dataset ([#1192](https://github.com/PyTorchLightning/pytorch-lightning/pull/1192))
 - Fixed running `on_validation_end` only on main process in DDP ([#1125](https://github.com/PyTorchLightning/pytorch-lightning/pull/1125))
+- Fixes `use_amp` issue ([#1145](https://github.com/PyTorchLightning/pytorch-lightning/pull/1145))
```
Comment (Borda) [inline CHANGELOG.md:84]:
```suggestion
```

**Pull request merged**